mdata/aggmetric.go

```diff
@@ -186,23 +186,34 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
 			if agg == nil {
 				return Result{}, fmt.Errorf("consolidator %q not configured", consolidator)
 			}
+
+			// note: only the raw AggMetric potentially has a ROB
+			// these children aggregator AggMetric never do, so result.Points is never set here
 			result, err := agg.Get(from, to)
 			if err != nil {
 				return Result{}, err
 			}
-			futurePoints := []schema.Point{}
 
 			a.RLock()
 			defer a.RUnlock()
 
 			if a.rob != nil {
-				futurePoints = a.rob.Get()
-				// we may need to update result.Oldest if there are no points in result.
-				if len(futurePoints) > 0 && len(result.Iters) == 0 && len(result.Points) == 0 {
-					result.Oldest = futurePoints[0].Ts
+				futurePoints := a.rob.Get()
+				if len(futurePoints) > 0 {
+
+					// There are 2 possibilities here:
+					// A) The AggMetric is new: it may have all its points in the ROB, and no points may have been flushed yet to the chunks (no result.Iters).
+					//    In this case, the oldest point comes from the ROB, so we update the Oldest field here.
+					// B) Points have been flushed by the aggregator into the chunks:
+					//    Because the aggregator (in Foresee but also in general) cannot accept points for boundaries that it has already flushed to the AggMetric's chunks,
+					//    we know that the oldest point will always come from the chunks and the field is already properly set.
+
+					if len(result.Iters) == 0 {
+						result.Oldest = futurePoints[0].Ts
+					}
+					result.Points = aggregator.Foresee(consolidator, from, to, futurePoints)
```

Comment:
I think we should call `Foresee` whether `len(futurePoints)>0` or not because it will flush the the current boundary interval regardless of whether it's complete or not.

Reply:
oops! you are correct. i just pushed the fix

Reply:
actually, we should probably stick to what you had orginally, which is even call Foresee when there isn't even a ROB, as a way to get the "work in progress" aggregation into the result. although that has nothing to do #1485 

It's a bit debatable whether such behavior is desired, but i think this is similar to queries like `sum(a,b)` when a point has come in for a, but not yet for b, this also leads to incomplete sums being returned. but that's how MT currently behaves, so i think it makes sense to extend that to the in-progress aggregation, although this is a breaking change, so it should go into the changelog. thoughts @shanson7 ?

Reply:
Yes that's debatable (I am also not sure which one is desired.)
* If we stick with the original one, it makes sense for `min/max/avg` but for `sum`, users will see the last aggregated point increases from the a small value gradually...
* If we choose the other one and if the rollup span is big enough like 1-day rollup, users will have to wait a day to see the latest result...

but for either case,

* if we don't want the last incomplete boundary we should fix Foresee to not aggregate on the future boundary and call it inside `if len(futurePoints) > 0`.
* Otherwise we should call it outside `if a.rob != nil` (2 lines below the current one)

I think that's why the UT failed.

Reply:
> It's a bit debatable whether such behavior is desired, but i think this is similar to queries like `sum(a,b)`

I think it also lines up with how runtime consolidation works. If we are using raw data, but consolidating 5 points together (e.g. using `sum`), we would see the most recent point slowly grow until it is "complete". I think roll-ups should behave similarly to runtime consolidation in that respect.

Reply:
I have just done a little experiment...
```
~/g/s/g/D/s/graphite-docker ❯❯❯ cat storage-aggregation.conf
[default]
pattern = .*
xFilesFactor = 0
aggregationMethod = sum
~/g/s/g/D/s/graphite-docker ❯❯❯ cat storage-schemas.conf
[default]
pattern = .*
retentions = 5s:1h,5m:6d

docker run -d \
 --name graphite \
 --restart=always \
 -p 80:80 \
 -p 2003-2004:2003-2004 \
 -p 2023-2024:2023-2024 \
 -p 8125:8125/udp \
 -p 8126:8126 \
 -v /home/dieter/go/src/github.com/Dieterbe/sandbox/graphite-docker/storage-schemas.conf:/opt/graphite/conf/storage-schemas.conf \
 -v /home/dieter/go/src/github.com/Dieterbe/sandbox/graphite-docker/storage-aggregation.conf:/opt/graphite/conf/storage-aggregation.conf \
 graphiteapp/graphite-statsd

./build/mt-fakemetrics feed --mpo 10 --period 5s --flush 5s --carbon-addr 'localhost:2003'                                                                                           ✘ 130 
2022/02/08 11:52:31 [I] starting listener on :6764
params: metricName=some.id.of.a.metric.%d, orgs=1, mpo=10, period=5, flush=5000, offset=0, speedup=1, stopAtNow=false
per org:         each 5s, flushing 10 metrics so rate of 2 Hz. (10 total unique series)
times    1 orgs: each 5s, flushing 10 metrics so rate of 2 Hz. (10 total unique series)
```

with this setup, when i continuously run `curl 'http://localhost/render/?target=some.id.of.a.metric.1&format=json' | jsonpp  | tail -n 14` i can see how the last point is always at the last 5min boundary that passed, and it gradually grows during the 5 min interval until it settles and the next point begins (reminder: graphite uses premarking). actually, graphite does something weird where sometimes if you query at exactly a 5s aligned time, sometimes the value is very close to 0, but that seems like a bug

so, it seems we agree with what graphite already does and the behavior @zlin179 had introduced.  i will re-instate that and add a note to the changelog.